starter_packs/views.py

```
@@ -671,3 +690,303 @@ def follow_bg(user_id: int, starter_pack_slug: str):
+
+    avatar = avatar.convert("RGBA")
+    avatar = avatar.resize((crop_mask.width, crop_mask.height), resample=Image.Resampling.LANCZOS)
+    modified = Image.new("RGBA", (crop_mask.width, crop_mask.height), (127, 127, 127, 255))
+    modified.alpha_composite(avatar)
+    modified.putalpha(crop_mask)
+    return modified
+
+
+def render_splash_image(starter_pack, host_attribution):
+    """
+    Renders (or re-renders) a splash image for a specific starter pack.
+    The result is returned as a PIL image and stored in the media
+    directory. Note that this triggers a download of up to
+    `SPLASH_IMAGE_NUMBER_OF_AVATARS` remote avatars and a good amount
+    of CPU and RAM use -- this operation is _expensive_! The caller is
+    responsible for sensible scheduling.
+
+    Note: `host_attribution` is cached eagerly by the background gradient
+    renderer. The value is only used if no background for the configured
+    render size is in the local cache.
+    """
+
+    resolution = (1200, 630)  # final rendered image size in pixels
+    supersampling_factor = 3  # scaling factor for render canvas size (to avoid aliasing)
+    media_dir = settings.BASE_DIR / "media"
+    splash_dir = settings.BASE_DIR / "media" / "splash"
+
+    for directory in (media_dir, splash_dir):
+        try:
+            directory.mkdir(mode=0o770, exist_ok=True)
+        except (FileExistsError, PermissionError) as e:
+            # With `exist_ok=True`, the FileExistsError is only raised if the media path
+            # already exists and is _not_ a directory.
+            logger.exception("Unable to create %s directory: %s", directory, type(e).__name__)
+            return
+        try:
+            write_test_path = directory / "__write_test.lock"
+            write_test_path.touch()
+            write_test_path.unlink()
+        except OSError:
+            logger.exception("Unable to write to directory %s", directory)
+            return
```

I think there is no need for trying to create/test dirs every time we run `render_splash_image`. I think Django will take care of dir creation for us and if it doesn't I can create them manually once and it should work from then on.

```suggestion
```

---

Django may create `media` if it is smart enough to deduce it from the settings, but it will definitely not create `media/splash` since that's only used programmatically here.

I did the write test at the start so we could fail fast if something is wrong with the directories, but if we remove the check here, what ends up happening is that we need to wrap the file writes at the ends of `render_splash_image` and `get_splash_background` in `try ... except` instead. I have no problem with that, the only downside is if there is a problem with the file system after all, we may end up downloading the avatars and generating the splash image before we realize we can't save the result.